WB-119: Make snapshots include all docs, not just root

somewhatabstract opened the pull request:
Jira: [WB-119](https://khanacademy.atlassian.net/browse/WB-119)

# Code Changes
## 1st commit
This should be reviewed.
[Review commit](https://github.com/Khan/wonder-blocks/pull/124/commits/c57d2c9c91673165cdfd4aed44e520a1bb2e3e7c)
This changes the snapshot test generation script so that it collates examples and file paths from all our docs. There is some extensive refactoring of the main script so that I could reuse various bits and keep methods easy to follow.

## 2nd commit
This should be reviewed.
[Review commit](https://github.com/Khan/wonder-blocks/pull/124/commits/c8280b4378ae7db4e962c8e6d271f584ec0d17cc)
This also then updates all the broken markdown that was referencing files from its own source so that we don't have to munge require paths to be relative from the right locations (styleguidist still works).

# Auto-generated Changes
## 3rd commit
This is autogenerated and probably shouldn't be reviewed in great detail.
[Review commit](https://github.com/Khan/wonder-blocks/pull/124/commits/f0483635b17964260a25ceb9aac8a1455507fabd)
We have generated new snapshot tests.

## 4th commit
This is autogenerated and doesn't need direct review.
[Review commit](https://github.com/Khan/wonder-blocks/pull/124/commits/e39f733a48d8075b93f2ca63c84bed6e9404e8ee)
We have generated new snapshots so that tests pass.

jeresig commented:
Deploy preview for *wonder-blocks* ready!

Built with commit dde278a0d083b62541c0c159297039e44d98b75a

https://deploy-preview-124--wonder-blocks.netlify.com

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -27,12 +27,13 @@ const ignored = [
         "// Do not edit this file.  To make changes to these snapshot tests",
-        `// edit ${root}/docs.md and run \`npm run gen-snapshot-tests\`.`,
+        `// edit ${root}/docs.md and other documentation files in the package,`,
+        "// then run `npm run gen-snapshot-tests`.",
```
Updated the comment to reflect what now happens (should probably update this to yarn too, but I don't know that I want to do all that rebasing and such. Maybe as a future edit).

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
-        continue;
+
+    return examples;
+}
```
Helper function so we can reuse it. All it does is grab examples. Also encapsulates checking that the path is valid, just for ease of use.

kevinbarabash replied:
Nice!

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+            readExamplesFromDocument(componentDoc),
+        );
+    }
+}
```
This was the content of the old `glob` callback. It now takes a collation object and reads examples from each components docs, if they exist. We only collate once since we don't want the examples repeated.

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+    }
+}
+
+function extractExamplesAndComponentFiles(section, collation) {
```
This function is the bulk of the old generation method.

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+    if (section.components) {
+        const componentGlobs = typeof section.components === "string"
+            ? [section.components]
+            : section.components;
```
Just normalizing the components value since it can be a single string or an array of strings.

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+            ? [section.components]
+            : section.components;
+        for (const componentGlob of componentGlobs) {
+            const files = glob.sync(componentGlob, {ignore: ignored});
```
Synchronous `glob` call instead of async

somewhatabstract commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
-            });
-        } else {
-            generateTestFile(root, examples);
+    if (section.sections) {
```
This is the new bit to support situations like how the modal docs use sections to specify examples.

kevinbarabash commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+    const src = fs.readFileSync(file, "utf8");
+    const match =
+        src.match(/export default class \s*(\w+)/) ||
+        src.match(/export default \s*\w+\(\s*(\w+)/);
```
This is to support things that aren't classes?

somewhatabstract replied:
Yeah, that was already there in the old code. It is grabbing things like our stuff that gets wrapped by MediaLayout (FixedCell, for example).

kevinbarabash replied:
right, I remember now... man my memory is bad 😅

kevinbarabash commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+    return examples;
+}
+
+function extractExamplesAndComponentsForFile(file, collation) {
```
Can you document what `file` should be passed in.  It looks a like a .js file that exports a component.

somewhatabstract replied:
Ah yes, I can add some docs.

kevinbarabash commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+}
+
+function extractExamplesAndComponentFiles(section, collation) {
+    // Add any examples fromt he section content.
```
`from the section`

somewhatabstract replied:
Great catch! My chubby hammy fisters must've messed that up.

kevinbarabash commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+        examples: [],
+        componentFileMap: {},
+    };
+    extractExamplesAndComponentFiles(section, collation);
```
Could `extractExamplesAndComponentFiles` return a `collation` instead of mutating the arguments.

somewhatabstract replied:
I had played around with that. We could easily change to that model, i think. I'll take a look.

kevinbarabash commented on utils/gen-snapshot-tests.js:
```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+        for (const childSection of section.sections) {
+            extractExamplesAndComponentFiles(
+                childSection,
+                collation,
```
This is awesome!  This should handle an level of nesting we do with sections and subsections.  Thanks for doing this.

somewhatabstract commented:
OK, @kevinbarabash, I have made some changes:

1. Fixed spelling in that comment
1. Updated the main generated file comment to refer to yarn and try to conform to our 80 char width commentary standard
1. Updated to return the collation object rather than mutating things in place
1. Merged with master and updated modal snapshots to cater for close button moving
1. Added docs

somewhatabstract commented:
Sorry for the commit mess around. Did a rebase when I should've done something else and got things all over the place. Anyway, it's all good now :smile:

kevinbarabash commented:
@somewhatabstract are these commits safe to commit as is or should I squash them?

somewhatabstract commented:
@kevinbarabash All good now. I would commit as-is that way we have the nice separation (which is what I was striving for but somehow got all mixed up). 

We are all good now though. Triple checked. So assuming everything passes CI and you like the changes, we're good to go.

somewhatabstract commented:
I've updated the links in the PR discussion to match the commits so that they can at least be found if we want to revisit this for some reason. :)

kevinbarabash commented:
I've merged them via rebasing.
